components/support/webextensions/src/main/java/mozilla/components/support/webextensions/WebExtensionSupport.kt

```diff
@@ -119,21 +133,50 @@ object WebExtensionSupport {
             }
 
             override fun onInstalled(webExtension: WebExtension) {
-                installedExtensions.add(webExtension)
-                store.dispatch(WebExtensionAction.InstallWebExtension(webExtension.toState()))
-
-                // Register action handler for all existing engine sessions on the new extension
-                store.state.tabs
-                    .forEach {
-                        val session = it.engineState.engineSession
-                        if (session != null) {
-                            registerSessionActionHandler(webExtension, session, SessionActionHandler(store, it.id))
-                        }
-                    }
+                registerInstalledExtension(store, webExtension)
             }
         })
     }
 
+    /**
+     * Awaits for completion of the initialization process (completes when the
+     * state of all installed extensions is known).
+     */
+    suspend fun awaitInitialization() = initializationResult.await()
+
+    /**
+     * Queries the [engine] for installed web extensions and adds them to the [store].
+     */
+    private fun registerInstalledExtensions(store: BrowserStore, engine: Engine) {
+        engine.listInstalledWebExtensions(
+            onSuccess = {
+                extensions -> extensions.forEach { registerInstalledExtension(store, it) }
+                initializationResult.complete(Unit)
+            },
+            onError = {
+                throwable -> logger.error("Failed to query installed extension", throwable)
+                initializationResult.completeExceptionally(throwable)
+            }
+        )
+    }
+
+    /**
+     * Marks the provided [webExtension] as installed by adding it to the [store].
+     */
+    private fun registerInstalledExtension(store: BrowserStore, webExtension: WebExtension) {
```

Comment:
fun overloading intended?

Reply:
@psymoon don't see any overloading here? There's a plural version (`registerInstalledExtensions`) which acts on all installed extensions and one that acts on just one (`registerInstalledExtension`).

Reply:
Ah, my mistake 👍